packages/shared/ReactLazyComponent.js

```diff
@@ -9,44 +9,66 @@
 
 export type Thenable<T, R> = {
   then(resolve: (T) => mixed, reject: (mixed) => mixed): R,
-  ...
 };
 
-export type LazyComponent<T> = {
+type UninitializedLazyComponent<T> = {
   $$typeof: Symbol | number,
-  _ctor: () => Thenable<{default: T, ...}, mixed>,
-  _status: 0 | 1 | 2,
-  _result: any,
-  ...
+  _status: -1,
+  _result: () => Thenable<{default: T, ...} | T, mixed>,
+};
+
+type PendingLazyComponent<T> = {
+  $$typeof: Symbol | number,
+  _status: 0,
+  _result: Thenable<{default: T, ...} | T, mixed>,
 };
 
 type ResolvedLazyComponent<T> = {
   $$typeof: Symbol | number,
-  _ctor: () => Thenable<{default: T, ...}, mixed>,
   _status: 1,
-  _result: any,
-  ...
+  _result: T,
 };
 
+type RejectedLazyComponent = {
+  $$typeof: Symbol | number,
+  _status: 2,
+  _result: mixed,
+};
+
+export type LazyComponent<T> =
+  | UninitializedLazyComponent<T>
+  | PendingLazyComponent<T>
+  | ResolvedLazyComponent<T>
+  | RejectedLazyComponent;
+
 export const Uninitialized = -1;
 export const Pending = 0;
 export const Resolved = 1;
 export const Rejected = 2;
 
 export function refineResolvedLazyComponent<T>(
   lazyComponent: LazyComponent<T>,
-): ResolvedLazyComponent<T> | null {
+): T | null {
   return lazyComponent._status === Resolved ? lazyComponent._result : null;
 }
 
 export function initializeLazyComponentType(
   lazyComponent: LazyComponent<any>,
 ): void {
   if (lazyComponent._status === Uninitialized) {
-    lazyComponent._status = Pending;
-    const ctor = lazyComponent._ctor;
+    let ctor = lazyComponent._result;
+    if (!ctor) {
```

Comment:
Coercion is OK here?

Reply:
When we detect existence of something unknown things we tend to do this type of check.